src/Control/Category/Cartesian.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE MonoLocalBinds #-}
+module Control.Category.Cartesian where
+
+import Control.Category (id, (>>>))
+import Data.Functor.Contravariant
+import Data.Void
+import Prelude hiding (id)
+
+import Control.Category.Tensor
+
+
+class Symmetric cat t => Semicartesian cat t where
+  split :: cat a (a `t` a)
+
+  fork :: cat a x -> cat a y -> cat a (x `t` y)
+  fork f g = split >>> f # g
+
+  infixr 9 /\
+  (/\) :: cat a x -> cat a y -> cat a (x `t` y)
+  (/\) = fork
+  {-# MINIMAL split #-}
+
+
+class Symmetric cat t => Semicocartesian cat t where
+  merge :: cat (a `t` a) a
+
+  fuse :: cat x a -> cat y a -> cat (x `t` y) a
+  fuse f g = f # g >>> merge
+
+  infixr 9 \/
+  (\/) :: cat x a -> cat y a -> cat (x `t` y) a
+  (\/) = fuse
+  {-# MINIMAL merge #-}
+
+
+instance Semicartesian (->) t => Semicocartesian Op t where
+  merge = Op split
+
+instance Semicocartesian (->) t => Semicartesian Op t where
+  split = Op merge
```

Comment:
> I would imagine that with the 2 classes we're writing here, we would write four instances: a Cartesian instance for (->) (,) (), a Cartesian instance for (<-) Either Void, a Cocartesian instance for (->) Either Void, and a Cocartesian instance for (<-) (,) (). Is that right? Or do we omit a pair of these?

we omit a pair of these via instances like this

Reply:
> The main issue I have with the approach here is that there's essentially a combinatorial explosion of duplication that occurs in stuff that we hang off Cartesian and Cocartesian. Any operations you define in terms of Cartesian won't work for Cocartesian (and thus must be redefined), and vice versa.

Is this also true with the above instances?

Reply:
Ah, I see. Well we are still defining the pair of instances though, no? The definitions are certainly very simple, but the (perhaps not very dire) problem I'm trying to indicate is that one tends to combinatorially accumulate such "trivial" definitions unless duality is exploited. There's an argument to be made that there is not much to define downstream of `Cartesian`, but you will have this kind of problem with e.g. different kinds of distributive structures (which we haven't got to yet).